### PR TITLE
Warrior primordial : Edition 2

### PR DESCRIPTION
**Pull request by Canohaaaaaaa**
## About The Pull Request
Adds Warrior's primordial

Primordial warriors gain jab : a ranged punch with a heavier slow. On top of that every 3rd ability cast by a primordial will be empowered. Only works during combat, so don't expect to START with an empowered ability.

empowered

### Empowered fling
Briefly knocks down every marine down the path of the flung victim. (basically bowling)
### Empowered grapple toss
Allows you to throw twice as far.
### Empowered jab
"Blinds" and confuses the victim for 3 seconds.
### Empowered punch
Does 50% more damage.
### Lunge
Lunge doesn't actually have an empowered version, instead it functions as a "wildcard" and can be used at the beginning or end of a combo since it counts as an offensive ability but doesn't consume the empowered cast!

Lunge also automatically selects grapple toss for you on a successful catch so you don't actually need a new keybind for jab.

The first iteration of this primordial was too divisive and didn't keep warrior's regular gameplay. This one adds a little depth to warrior's gameplay on primordial, making them think about how to string their abilities for maximum effectiveness.

## Why It's Good For The Game

Something new to spend psy points on!

### CODE REVIEW

**Comment by vilereaver41** [comment]
No sucker punch?

**Comment by InterroLouis** [comment]
Fuck no

**Comment by ilikesaying** [comment]
literally CM boxer smh

**Comment by ThelastArchiver** [comment]
warriors be like " MUDA MUDA MUDA MUDA MUDA MUDA MUDA MUDA MUDA"

**Comment by Mrrpip** [comment]
I mean....What if a player preffered lunge and fling?

**Comment by Canohaaaaaaa** [comment]
> 
> 
> I mean....What if a player preffered lunge and fling?

I know it's a bit weird for this premordial to be a "conversion" rather than a strict upgrade but I remember being told that an ability should be ditched on this one, which one exactly I forgot so i'll wait for tivi to review first.

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:366]: whats up with chan

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:523]: reccomend you use REF() instead of name as names can be identical

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:518]: please use more descriptive var names, A is usually used for declaring atoms

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:521]: no need to get a ref to the combo list here, just use uppercut_ability.combo

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:531]:
```suggestion
	addtimer(CALLBACK(uppercut_ability, /datum/action/xeno_action/activable/uppercut.proc/clear_combo, A, counter), WARRIOR_COMBO_FADEOUT_TIME, TIMER_OVERRIDE|TIMER_UNIQUE)
```

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:534]: del() usage is not allowed as it is extremely slow
instead please just remove all references to the image instead and it will get garbage collected

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:621]:
```suggestion
/datum/action/xeno_action/activable/uppercut/can_use_ability(atom/A, silent = FALSE, override_flags)
	. = ..()
	if(!.)
		return
```

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:618]:
```suggestion
	keybind_signal = COMSIG_XENOABILITY_UPPERCUT
	/// Who did we punch/jab recently and how many times ?
	var/list/combo
```

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:652]: todo list?

Also as a side note I'd like to discuss these stuns @ me in discord

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:639]: snake_case instead of CamelCase please

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:320]: Autodoc new vars please, see https://discord.com/channels/498569646014857217/500801682058379265/775809656656101396 for how to do it

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Well done for a first pr

**Comment by Al-1ce** [comment]
I am so fucking tired of this word. Primordial this, primordial that. Jesus. Christ. Can't you think of something original for once?

**Comment by Lumipharon** [comment]
I feel like this isn't really an upgrade for warriors, considering how CM is a fair bit different in how it plays compared to TGMC.

While it will likely be great for 1v1 or small group situations, outside of that you lose your control abilities for stuns/damage that have very high risk but honestly not great rewards considering plural marines can be blasting away at you with aimmed fire before you can get enough combo up.

Then even if you do wombocombo someone, unless you can secure the body 9/10 times they get dragged off and revived very easily.
As a normal warrior, being able to lunge and grapple throw means the odds of truly lasting damage (i.e. killing them and holding onto the body to be sucked if not perma'd) is quite good.

I could be wrong though.

**Comment by Canohaaaaaaa** [comment]
Alright, design lead weighted in. I'm going to keep warrior's default abilties jab (albeit without the reset) and the stacking mechanic but remove uppercut, I'll figure out what to do with the stacks but for now they'll for sure enhance your punches (and maybe your regular abilities).
If you have a great idea @Nitrome#6861 on discord.

**Comment by DeltaFire15** [comment]
> Lunge also automatically selects grapple toss for you on a successful catch so you don't actually need a new keybind for jab.

This seems kinda weird since Warriors chain lunge into punch and not grapple toss around half of the time or so and this *might* interfere with manual ability switching slightly (I don't remember if mid-lunge ability selecting works)

**Comment by Canohaaaaaaa** [comment]
Grapple toss can only be used after a successful lunge (since you're realistically not throwing anyone with a regular grab) so it feels only natural for it to be the default selected ability after a lunge, regarding the lunge+punch it mechanically changes nothing since you can still lunge and press your punch bind for it to switch from toss if you wish to get the bonus damage.

**Comment by DeltaFire15** [comment]
> Grapple toss can only be used after a successful lunge (since you're realistically not throwing anyone with a regular grab) so it feels only natural for it to be the default selected ability after a lunge, regarding the lunge+punch it mechanically changes nothing since you can still lunge and press your punch bind for it to switch from toss if you wish to get the bonus damage.

Yup, see, the issue just is that prior to this, you can select your chosen follow-up ability mid-lunge to execute once you land and can click your target. This interferes with that by hard-selecting lunge after you land, even if you had punch readied during the lunge which slightly interferes with clean comboing since you have to be sure you landed before selecting punch, or you might "misfire" toss them onto their own tile instead which is pretty bad.

**Comment by Canohaaaaaaa** [comment]
fair point, i'll fix that later.

**Review by Kuro020 — APPROVED** [review]

File: code/datums/actions/action.dm
Comment (BraveMole) [inline code/datums/actions/action.dm:87]: autodoc (aka ///comment)

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:109]: bad indent

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:187]: abilities should be portable on other xenos without too much trouble, so no cast

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:91]: Should be a proc on xenomorph, with this code in the ovveridden version for warrior

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:206]: useless var, just call X.empower() directly?

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:212]: early return :
if(!empowered)
   continue

quick cast:
for(var/mob/living/carbon/human/human in temp)

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:282]: ditto, no cast should be needed

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:101]: autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:113]: autodoc

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:106]:
```suggestion
		for(var/datum/action/xeno_action/A AS in actions)
```

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:102]: you checked that already in empower

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:105]: early return here too

**Review by BraveMole — CHANGES_REQUESTED** [review]
Good job, the code is really neat

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (Canohaaaaaaa) [reply code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:206]: Can't exactly do that since the check for empowered happens in a loop, calling it on each one would give lots of combo.

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm
Comment (Canohaaaaaaa) [reply code/modules/mob/living/carbon/xenomorph/castes/warrior/warrior.dm:102]: there is a call for combo without empowered in lunge, since I don't want it to consume combo but still give some. Hence the check.

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:206]: hum right

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:213]: still need the quick cast, for var/mob/living/carbon/human in temp)

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:475]: fix this please, this should be a overwrirtten proc on mob/living/carbon, rather that a shitty istype(src)

File: code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/warrior/abilities_warrior.dm:107]: need a null check, what if you are giving lunge to another xeno.

grapple_toss?.action_activate

**Review by BraveMole — CHANGES_REQUESTED** [review]

**Comment by Lumipharon** [comment]
Lunge + punch + (empowered)fling sounds hilarious now.
Looks good.

**Comment by Luxyboo** [comment]
as worrier main i really dislike the automatic switch from lunge to toss, i would prefer it to be an optional thing or not in at all. just my thoughts though.

**Comment by Canohaaaaaaa** [comment]
> as worrier main i really dislike the automatic switch from lunge to toss, i would prefer it to be an optional thing or not in at all. just my thoughts though.

I played around with it a bit myself, turns out miss inputs were making me instantly toss them on the tile I lunged at them from. 
Was it what happened to you ? If so I was thinking about adding a slight delay to prevent that from happening.
Or was it because of muscle memory and the fact you're too used to pressing the toss key ?

I would like a bit more input so I can tune it, if the only solution is a crappy compromise i'll scrap the feature. It will mean you'll have a massive amount of binds for warrior though.

**Comment by BraveMole** [comment]
Looks ok, tell me if you want more test or merge

**Comment by Texicos** [comment]
Canoh, I have found out that you can lunge, jab the downed marine without the lunge breaking, and then do an empowered punch with the lunge grab damage bonus to the punch as well. This means a good 90 damage is dealt from the one punch. Also Jab is effected by a 50% damage bonus from the lunge hold. this means you can deal 115-118 damage in a single combo. Plz fix men

**Comment by Luxyboo** [comment]
> > as worrier main i really dislike the automatic switch from lunge to toss, i would prefer it to be an optional thing or not in at all. just my thoughts though.
> 
> I played around with it a bit myself, turns out miss inputs were making me instantly toss them on the tile I lunged at them from. Was it what happened to you ? If so I was thinking about adding a slight delay to prevent that from happening. Or was it because of muscle memory and the fact you're too used to pressing the toss key ?
> 
> I would like a bit more input so I can tune it, if the only solution is a crappy compromise i'll scrap the feature. It will mean you'll have a massive amount of binds for warrior though.

i think the feature is good i just feel like it will really mess with muscle memory for people like me who do lunge then punch instead of lunge then toss

i was hoping you could do something to where you would have to enable the switch to toss

**Comment by DeltaFire15** [comment]
From having played a bit of Warrior with the testmerge, yeah it does mess with muscle memory a bit as clicking after lunging is now unsafe (aka no more spamclicking or at least gotta react faster) 
Did cause a few misthrows due to that.
Then again, I don't know how other Warriors do it with their lunges.

**Review by BraveMole — APPROVED** [review]
Looks good to me, fix conflicts and merge

**Comment by Luxyboo** [comment]
also i have a bug to report: attempting to punch a missing limb will not consume plasma or activate a cooldown but it will count towards empowerment

**Comment by Mechanelize** [comment]
> I would like a bit more input so I can tune it, if the only solution is a crappy compromise i'll scrap the feature. It will mean you'll have a massive amount of binds for warrior though.

Please scrap the auto switch already, it fucks with muscle memory and adds a longer delay to fling or punch after a good lunge.

**Comment by DrCrawler** [comment]
Please man PLEASE make the autotoss toggleable with a hud button or atleast in prefs, this is turning form QoL into torture due to my muscle memmory and spamclicking so i actualy grapple something.

**Comment by Canohaaaaaaa** [comment]
Scrapped auto-swap, requires more work so I might make a proper PR for it in the future.
Nerfed full damage combo potential down from 115 -> 92. Jab forces you to drop whoever you're pulling just like punch now.

**Review by BraveMole — APPROVED** [review]

**Pull request merged**
